queue/driver_mongo.go

```diff
@@ -634,35 +623,28 @@ func (d *mongoDriver) SaveAndPut(ctx context.Context, toSave amboy.Job, toPut am
 		return nil, nil
 	}
 
-	if _, err = sess.WithTransaction(ctx, atomicSaveAndPut); err != nil {
-		return errors.Wrap(err, "swapping scopes")
+	if _, err = sess.WithTransaction(ctx, atomicCompleteAndPut); err != nil {
+		return errors.Wrap(err, "atomic complete and put")
 	}
 
 	return nil
 }
 
 func (d *mongoDriver) Complete(ctx context.Context, j amboy.Job) error {
-	job, err := d.prepareInterchange(j)
+	ji, err := d.prepareInterchange(j)
 	if err != nil {
 		return errors.WithStack(err)
 	}
 
-	isRetryableJob := amboy.WithRetryableJob(j, func(rj amboy.RetryableJob) {
-		info := rj.RetryInfo()
-		if info.Retryable && info.NeedsRetry && rj.ShouldApplyScopesOnEnqueue() {
-			// If the job is supposed to retry and apply the scopes immediately
-			// to the retry job, we cannot let go of the scopes yet because they
-			// will need to be safely transferred to the retry job.
-			return
-		}
-		job.Scopes = nil
-	})
-
-	if !isRetryableJob {
-		job.Scopes = nil
+	// It is safe to drop the scopes now in all cases except for one - if the
+	// job still needs to retry and applies its scopes immediately to the retry
+	// job, we cannot let go of the scopes yet because they will need to be
+	// safely transferred to the retry job.
+	if !ji.RetryInfo.Retryable || !ji.RetryInfo.NeedsRetry || !ji.ApplyScopesOnEnqueue {
```

Comment:
For scopes applied on enqueue, the hand-off has to occur in the retry handler, so it's deferring the scope lock drop until `CompleteAndPut` occurs.

This logic is identical to what it was before, but I applied DeMorgan's law to make the conditional logic less awkward-looking. It's equivalent to this:
```go
if ji.RetryInfo.Retryable && ji.RetryInfo.NeedsRetry && ji.ApplyScopesOnEnqueue {
    // pass, hold the scope
} else {
    ji.Scopes = nil
 }
```

Reply:
Maybe I am missing something, but I do not see where the scope hand-off is happening in the retry handler. I am assuming it would be in `tryEnqueueJob` after `CompleteAndPut` is called. Could you point me to it?

Reply:
It's implicit by calling `GetAttempt()` since it receives an in-memory copy of the scopes it should acquire. Dropping the scopes here by settings `NeedsRetry = false` is the only step necessary for the new job to acquire them in `Put()` (L531).

Reply:
Ah ok, got it.